network_simulator/statistics.py

```diff
@@ -4,6 +4,186 @@
+    to when a packet was transmitted.
+    """
+    def __init__(self, buffer_occupancy=0, packet_loss_times=[],
+                 packet_transmit_times=[]):
+        self.buffer_occupancy = buffer_occupancy
+        self.packet_loss_times = packet_loss_times
+        self.packet_transmit_times = packet_transmit_times
+
+    # TODO(sharon): functions and subclasses/events (if any) goes here.
+
+
+class FlowStats(object):
+    """
+    Contains information of a flow regarding to per-flow send/receive rate
+    and packet round-trip delay.
+
+    :ivar list packet_sent_times_sec: a list of timestamps corresponding to
+    when a data packet was sent.
+    :ivar list packet_rec_times_sec: a list of timestamps corresponding to
+    when an ACK packet was received for a given :class:`.Flow`.
+    :ivar list packet_rtts_sec: a list of RTTs (in seconds).
+    """
+    def __init__(self, packet_sent_times_sec=[], packet_rec_times_sec=[],
+                 packet_rtts_sec=[]):
+        packet_sent_times_sec = packet_sent_times_sec
+        packet_rec_times_sec = packet_rec_times_sec
+        packet_rtts_sec = packet_rtts_sec
+
+    # TODO(sharon): functions and subclasses/events (if any) goes here.
+
+
+class HostStats(object):
+    """
+     Contains information of a host regarding to per-host send/receive rate.
+
+    :ivar list packet_sent_times_sec: a list of (timestamp, packet_size)
+    tuples.
+    :ivar list packet_rec_times_sec: a list of (timestamp, packet_size)
+    tuples.
+    """
+    def __init__(self, packet_sent_times_sec=[], packet_rec_times_sec=[]):
+        packet_sent_times_sec = packet_sent_times_sec
+        packet_rec_times_sec = packet_rec_times_sec
+
+    # TODO(sharon): functions and subclasses/events (if any) goes here.
+
+
 class Statistics(object):
-    # TODO(sharon): implement.
-    pass
\ No newline at end of file
+    """
+    Intended to be owned by the main loop to record all statistics within
+    the network.
+
+    :ivar dict link_stats: a map from (end_1_id, end_2_id) to
+    :class:`.LinkStats`.
+    :ivar dict flow_stats: a map from flow ID to :class:`.FlowStats`.
+    :ivar dict host_stats: a map from host ID to :class:`.HostStats`.
+    """
+    def __init__(self, link_stats={}, flow_stats={}, host_stats={}):
+        self.link_stats = link_stats
+        self.flow_stats = flow_stats
+        self.host_stats = host_stats
+
+    def get_link_stats(self, link):
+        """
+        Check whether a link has already existed in the dictionary; if so,
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Link link: link to be determined.
+        """
+        # Create a tuple of the pair and check if it exists in the dict.
+        tup = (link.end_1_addr, link.end_2_addr)
+
+        if tup not in self.link_stats:
+            self.link_stats[tup] = LinkStats
```

Comment:
Should be LinkStats() since an instance is being constructed. Same for FlowStats and HostStats below.